### PR TITLE
MWPW-115213 - Uber App Recommender a.k.a. Quiz

### PR DESCRIPTION
**Pull request by sabyamon**
* Creating Uber App Recommender a.k.a. Quiz.
* Data Model - https://wiki.corp.adobe.com/pages/viewpage.action?spaceKey=WEBT&title=UAR+Excel+Data+Model

Resolves: https://jira.corp.adobe.com/browse/MWPW-115213

**Test URLs:**
- Before: https://main--milo--adobecom.hlx.page/drafts/xiasun/quiz-2/?martech=off
- After: https://uar-integration--milo--adobecom.hlx.page/drafts/xiasun/quiz-2/?martech=off

### CODE REVIEW

**Comment by mokimo** [comment]
Make sure to add all the involved devs to https://github.com/adobecom/milo/blob/main/CODEOWNERS

**Comment by echampio-at-adobe** [comment]
> Make sure to add all the involved devs to https://github.com/adobecom/milo/blob/main/CODEOWNERS

Thanks for the reminder. I add our names in this PR: https://github.com/adobecom/milo/pull/1280/files

File: libs/blocks/quiz-results/quiz-results.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
+import { createTag } from '../../utils/utils.js';
+import { getMetadata, handleStyle } from '../section-metadata/section-metadata.js';
+
+async function loadFragments(el, experiences) {
+  // eslint-disable-next-line no-restricted-syntax
+  for (const href of experiences) {
+    /* eslint-disable no-await-in-loop */
+    const a = createTag('a', { href });
+    el.append(a);
+    const { default: createFragment } = await import('../fragment/fragment.js');
```
Comment (chrischrischris) [inline libs/blocks/quiz-results/quiz-results.js:10]: nit: move this out of the for loop

File: libs/blocks/quiz-results/quiz-results.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  window.alloy_all ??= {};
+  window.alloy_all.data ??= {};
+  window.alloy_all.data._adobe_corpnew ??= {};
+  window.alloy_all.data._adobe_corpnew.digitalData ??= {};
+  window.alloy_all.data._adobe_corpnew.digitalData.page ??= {};
+  window.alloy_all.data._adobe_corpnew.digitalData.page.pageInfo ??= {};
+  window.alloy_all.data._adobe_corpnew.digitalData.page.pageInfo.customHash = hashValue;
+  if (debug === 'quiz-results') {
+    // eslint-disable-next-line no-console
+    console.log(`Setting a custom hash for pageload to: ${window.alloy_all.data._adobe_corpnew.digitalData.page.pageInfo.customHash}`);
```
Comment (chrischrischris) [inline libs/blocks/quiz-results/quiz-results.js:40]:
```suggestion
    console.log('Setting a custom hash for pageload to: ', hashValue);
```

File: libs/blocks/quiz-results/quiz-results.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  }
+}
+
+export const loadingErrorText = 'Could not load quiz results:';
```
Comment (chrischrischris) [inline libs/blocks/quiz-results/quiz-results.js:44]: super nit (can leave as is if you want) - string constants are usually all caps and declared at top of file
```suggestion
export const LOADING_ERROR = 'Could not load quiz results:';
```

File: libs/blocks/quiz-results/quiz-results.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  debug ??= params.get('debug');
+
+  el.replaceChildren();
+
+  let results = localStorage.getItem(localStoreKey);
+  if (!results) {
+    redirectPage(quizUrl, debug, `${loadingErrorText} local storage missing`);
+    return;
+  }
+  results = JSON.parse(results);
```
Comment (chrischrischris) [inline libs/blocks/quiz-results/quiz-results.js:66]: This will SyntaxError on invalid JSON

File: libs/blocks/quiz/quiz.js [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+  const [currentStep, setCurrentStep] = useState(0);
+  const [totalSteps, setTotalSteps] = useState(3);
+  const [prevStepIndicator, setPrevStepIndicator] = useState([]);
+  const [nextQuizViewsExist, setNextQuizViewsExist] = useState(true);
+  const [btnAnalytics, setBtnAnalytics] = useState(null);
+  const [btnClicked, setBtnClicked] = useState(false);
+  const initialUrlParams = getUrlParams();
+  const [urlParam, setUrlParam] = useState(initialUrlParams);
+  const VALID_QUESTIONS = useMemo(() => [], []);
+  const KNOWN_PARAMS = useMemo(() => ['martech', 'milolibs', 'quiz-data'], []);
```
Comment (chrischrischris) [inline libs/blocks/quiz/quiz.js:44]: Why are these all caps?

File: libs/blocks/quiz/quiz.js [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+  quizSections.append(a);
+  const { default: createFragment } = await import('../fragment/fragment.js');
+  await createFragment(a);
+}
+
+const App = ({
+  initialIsDataLoaded = false,
+  preQuestions = {}, initialStrings = {},
+}) => {
+  const [questionData, setQuestionData] = useState(preQuestions.questionData || {});
```
Comment (chrischrischris) [inline libs/blocks/quiz/quiz.js:25]: super nit: alpha sort all the useState vars

File: libs/blocks/quiz/quiz.js [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+
+      setStringData(dataStrings);
+      setQuestionData(questions);
+      setStringQList(strMap);
+      setQuestionList(qMap);
+
+      // wait for data to load
+      setDataLoaded(true);
+
+      // add quiz class to page
```
Comment (chrischrischris) [inline libs/blocks/quiz/quiz.js:70]: No need for the comments (and I think that `// wait for data to load` isn't actually what that's doing).

File: libs/blocks/quiz/quiz.js [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+      document.body.classList.add('quiz-page');
+    })();
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [setQuestionData, setStringData, setStringQList, setQuestionList]);
+
+  useEffect(() => {
+    function handlePopState() {
+      const currentURL = window.location.href;
+      window.location.href = currentURL;
```
Comment (chrischrischris) [inline libs/blocks/quiz/quiz.js:79]: nit - also the function name doesn't seem right - should it just be something like `reloadPage`?
```suggestion
      window.location.reload()
```

File: libs/blocks/quiz/quiz.js [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+    }
+  }, [userSelection, nextQuizViewsExist]);
+
+  /**
+   *  Updates the url param when user selects the options.
+   *  Happens with each option click/tap.
+   */
+  useEffect(() => {
+    const cardValues = Object.getOwnPropertyNames(selectedCards);
+    if (selectedQuestion) {
```
Comment (chrischrischris) [inline libs/blocks/quiz/quiz.js:126]: To reduce nesting early return

File: libs/blocks/quiz/quiz.js [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+      setUrlParam((prevUrlParam) => {
+        const newParam = { ...prevUrlParam };
+        if (selectedQuestion && cardValues.length === 0) {
```
Comment (chrischrischris) [inline libs/blocks/quiz/quiz.js:129]: May want to assign `selectedQuestion.questions` to a const with shorter name (`questions`?)

File: libs/blocks/quiz/quiz.js [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+  useLayoutEffect(() => {
+    if (Object.keys(urlParam).length > 0 && btnClicked === true) {
+      let urlParamList = Object.keys(urlParam).map((key) => {
+        const paramList = [...urlParam[key]];
+        if (paramList.length) {
+          return `${key}=${paramList.join(',')}`;
+        }
+        return null; // Explicitly return null if the condition is not met
+      }).filter((item) => !!item && !KNOWN_PARAMS.includes(item.split('=')[0]));
+      const knownParamsList = KNOWN_PARAMS.filter((key) => key in urlParam).map((key) => `${key}=${urlParam[key].join(',')}`);
```
Comment (chrischrischris) [inline libs/blocks/quiz/quiz.js:156]: nit: break into multiple lines

File: libs/blocks/quiz/quiz.js [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+
+    setSelectedCards(newState);
+    setCountOfSelectedCards(Object.keys(newState).length);
+  };
+
+  useEffect(() => {
+    const getStringValue = (propName) => {
+      if (!selectedQuestion) return '';
+      const question = stringQList[selectedQuestion.questions];
+      return question ? question[propName] || '' : '';
```
Comment (chrischrischris) [inline libs/blocks/quiz/quiz.js:263]: nit
```suggestion
      return question?.[propName] || '';
```

File: libs/blocks/quiz/utils.js [hunk 13]
```diff
@@ -0,0 +1,441 @@
...
+let configPath; let quizKey; let analyticsType; let analyticsQuiz; let metaData;
+
+const initConfigPath = (quizMetaData) => {
+  const quizConfigPath = quizMetaData.quizurl.text.toLowerCase();
+  const urlParams = new URLSearchParams(window.location.search);
+  const stringsPath = urlParams.get('quiz-data');
+  if (stringsPath) {
+    return (filepath) => `${stringsPath}/${filepath}`;
+  }
+  return (filepath) => `${quizConfigPath}${filepath}`;
```
Comment (chrischrischris) [inline libs/blocks/quiz/utils.js:18]: nit
```suggestion
  return (filepath) => `${stringsPath || quizConfigPath}${filepath}`;
```

File: libs/blocks/quiz/utils.js [hunk 14]
```diff
@@ -0,0 +1,441 @@
...
+  umbrellaProduct,
+) => {
+  const structureFragments = [];
+  structureFragsArray.forEach((frag) => {
+    const fragment = frag.trim();
+    resultResources?.data?.forEach((row) => {
+      if (umbrellaProduct) {
+        if (umbrellaProduct && row.product === umbrellaProduct) {
+          structureFragments.push(row[fragment]);
+        }
```
Comment (chrischrischris) [inline libs/blocks/quiz/utils.js:157]:
```suggestion
        if (umbrellaProduct && row.product === umbrellaProduct) {
          structureFragments.push(row[fragment]);
        } else ...
```

File: libs/blocks/quiz/utils.js [hunk 15]
```diff
@@ -0,0 +1,441 @@
...
+      } else if ((primaryProducts.length > 0 && primaryProducts.includes(row.product))) {
+        if (row[fragment]) {
+          structureFragments.push(row[fragment]);
+        }
+      }
```
Comment (chrischrischris) [inline libs/blocks/quiz/utils.js:162]:
```suggestion
      } else if ((primaryProducts.length > 0 && primaryProducts.includes(row.product)) && row[fragment]) {
          structureFragments.push(row[fragment]);
      }
```

File: libs/blocks/quiz/utils.js [hunk 16]
```diff
@@ -0,0 +1,441 @@
...
+      if (product && product === row?.product) {
+        insertFragment();
+      }
+
+      function insertFragment() {
+        if (row[fragKey]) {
+          row[fragKey]?.split(',').forEach((val) => {
+            fragArray.push(val.trim());
+          });
+        }
```
Comment (chrischrischris) [inline libs/blocks/quiz/utils.js:225]: nit
```suggestion
          row[fragKey]?.split(',').forEach((val) => {
            fragArray.push(val.trim());
          });
```

File: libs/blocks/quiz/utils.js [hunk 17]
```diff
@@ -0,0 +1,441 @@
...
+  rObj.filteredResults = filteredResults;
+  rObj.resultResources = results['result-fragments'];
+
+  return rObj;
+};
+
+const getRecommendedResults = (selectedDestination, defaultValue) => (selectedDestination.length
+  ? selectedDestination : defaultValue);
+
+// TODO: needs refactoring - can split to smaller functions
```
Comment (chrischrischris) [inline libs/blocks/quiz/utils.js:296]: Remove comment.  Create an issue to track it if it's important that you refactor it.

File: libs/blocks/quiz/utils.js [hunk 18]
```diff
@@ -0,0 +1,441 @@
...
+export const findMatchForSelections = (results, selections) => {
+  const recommendations = [];
+  const matchResults = [];
+  const defaultResult = [];
+
+  results.forEach((destination) => {
+    if (destination.result.indexOf('(') === -1) {
```
Comment (chrischrischris) [inline libs/blocks/quiz/utils.js:303]: nit
```suggestion
    if (!destination.result.includes('(')) {
```

File: libs/blocks/quiz/utils.js [hunk 19]
```diff
@@ -0,0 +1,441 @@
...
+    if (destination.result === 'default') {
+      defaultResult.push(destination);
+    }
+  });
+
+  // direct match ac and express in results destination. Applying and condition
+  const isProductsMatched = selections.primary.every((product) => matchResults.includes(product));
+
+  if (isProductsMatched) {
+    // lr, ai
```
Comment (chrischrischris) [inline libs/blocks/quiz/utils.js:315]: remove comment

File: libs/blocks/quiz/utils.js [hunk 20]
```diff
@@ -0,0 +1,441 @@
...
+        const flowStepsList = next.split(',');
+        // RESET the queue and add only the next question.
+        if (flowStepsList.includes('RESET')) { // Reset to intial question
+          nextQuizViews = []; // Resetting the nextQuizViews
+          // eslint-disable-next-line no-param-reassign
+          userFlow = []; // Resetting the userFlow as well
+          lastStopValue = 'RESET';
+        }
+
+        if (!hasResultTigger) {
```
Comment (chrischrischris) [inline libs/blocks/quiz/utils.js:387]: Tigger?

**Review by chrischrischris — COMMENTED** [review]
Nice!!!!  Just a handful of nits

File: libs/blocks/quiz-results/quiz-results.js (hunk 3 above)
Comment (colloyd) [reply libs/blocks/quiz-results/quiz-results.js:44]: This was put in place in service of unit tests to reduce repeated strings there. In that context, I opted not to go with all caps and I'd need to update the references to it in the test as well. If it's not a big deal, I'd prefer to leave it as is.

File: libs/blocks/quiz-results/quiz-results.css [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+@media screen and (min-width: 600px) {
+  .quiz-results .tablet-portrait-up {
+    display: block;
+  }
+
+  .quiz-results .phone-only,
+  .quiz-results .tablet-landscape-up,
+  .quiz-results .desktop-up, .quiz-results .big-desktop-up {
```
Comment (rgclayton) [inline libs/blocks/quiz-results/quiz-results.css:23]: nit. New line after the comma

File: libs/blocks/quiz-results/quiz-results.css [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+.quiz-results {
+  visibility: hidden;
+}
+
+.quiz-results.show {
+  visibility: unset;
+}
+
+.quiz-results .tablet-portrait-up, .quiz-results .tablet-landscape-up,
+.quiz-results .section.desktop-up, .quiz-results .big-desktop-up {
```
Comment (rgclayton) [inline libs/blocks/quiz-results/quiz-results.css:12]: nit. New line after the comma

File: libs/blocks/quiz-results/quiz-results.css [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+  }
+}
+
+@media screen and (min-width: 900px) {
+  .quiz-results .tablet-landscape-up {
+    display: block;
+  }
+
+  .quiz-results .tablet-portrait-up,
+  .quiz-results .desktop-up, .quiz-results .big-desktop-up {
```
Comment (rgclayton) [inline libs/blocks/quiz-results/quiz-results.css:34]: nit. New line after the comma

File: libs/blocks/quiz-results/quiz-results.css [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+    display: none;
+  }
+}
+
+@media screen and (min-width: 1200px) {
+  .quiz-results .desktop-up {
+    display: block;
+  }
+
+  .quiz-results .tablet-portrait-up, .quiz-results .tablet-landscape-up,
```
Comment (rgclayton) [inline libs/blocks/quiz-results/quiz-results.css:44]: nit. New line after the comma

File: libs/blocks/quiz-results/quiz-results.css [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+    display: none;
+  }
+}
+
+@media screen and (min-width: 1440px) {
+  .quiz-results .big-desktop-up {
+    display: block;
+  }
+
+  .quiz-results .tablet-portrait-up, .quiz-results .tablet-landscape-up,
```
Comment (rgclayton) [inline libs/blocks/quiz-results/quiz-results.css:55]: nit. New line after the comma

File: libs/blocks/quiz-results/quiz-results.js [hunk 26]
```diff
@@ -0,0 +1,101 @@
+import { createTag } from '../../utils/utils.js';
+import { getMetadata, handleStyle } from '../section-metadata/section-metadata.js';
+
+export const LOADING_ERROR = 'Could not load quiz results:';
```
Comment (rgclayton) [inline libs/blocks/quiz-results/quiz-results.js:4]: Will this need to be localized?

File: libs/blocks/quiz/quiz.css [hunk 27]
```diff
@@ -0,0 +1,377 @@
+.quiz-page * {
+  box-sizing: border-box;
+}
+
+.quiz-page .feds-topnav-wrapper {
+  padding: 0 20px;
```
Comment (rgclayton) [inline libs/blocks/quiz/quiz.css:6]: Why are you trying to adjust gnav for quiz pages?

File: libs/blocks/quiz/quiz.js [hunk 28]
```diff
@@ -0,0 +1,334 @@
...
+        window.removeEventListener('popstate', handlePopState);
+      };
+    }
+    return () => {};
+  }, [knownParams, isDataLoaded]);
+
+  useEffect(() => {
+    if (userFlow.length) {
+      const currentFlow = userFlow.shift();
+      if (!currentFlow?.length) {
```
Comment (rgclayton) [inline libs/blocks/quiz/quiz.js:88]: Do you need to check if there is a length if currentFlow is null?

File: libs/blocks/quiz/quiz.css (hunk 27 above)
Comment (echampio-at-adobe) [reply libs/blocks/quiz/quiz.css:6]: Hi @rgclayton,

Here's a thread between Rares and Cody in an earlier PR. Without it, our customized nav has the logo and sign-in hugging the edge uncomfortably. The quiz pages normally use a very minimalistic nav. 

https://github.com/adobecom/milo/pull/1219#discussion_r1317520632

File: libs/blocks/quiz-results/quiz-results.js (hunk 26 above)
Comment (echampio-at-adobe) [reply libs/blocks/quiz-results/quiz-results.js:4]: Hi @rgclayton. We do have a future story to localize the thing. https://jira.corp.adobe.com/browse/MWPW-136320

Comment (echampio-at-adobe) [reply libs/blocks/quiz-results/quiz-results.js:4]: Actually, no, not this particular message. This message is just useful when the debug flag is set. If there's a problem in quiz-results, we redirect to the uber creative cloud landing page anyway, and log a message in lana. The story I mentioned refers to other messages in the quiz block, not quiz-results.

File: libs/blocks/quiz-results/quiz-results.css (hunk 22 above)
Comment (echampio-at-adobe) [reply libs/blocks/quiz-results/quiz-results.css]: Commas should be in the appropriate places in CSS now.

File: libs/blocks/quiz-results/quiz-results.css (hunk 24 above)
Comment (echampio-at-adobe) [reply libs/blocks/quiz-results/quiz-results.css]: Commas should be in the appropriate places in CSS now.

File: libs/blocks/quiz/quiz.js (hunk 28 above)
Comment (echampio-at-adobe) [reply libs/blocks/quiz/quiz.js]: I updated the code to test both userFlow and currentFlow in their respective locations before looking within them.

File: libs/blocks/quiz-results/quiz-results.css (hunk 23 above)
Comment (echampio-at-adobe) [reply libs/blocks/quiz-results/quiz-results.css]: Commas should be in the appropriate places in CSS now.

**Comment by echampio-at-adobe** [comment]
Hi @chrischrischris and @rgclayton. I believe we've addressed all of your comments now. Thanks for taking a look, and thanks in advance for a hopeful, final approval.

**Review by echampio-at-adobe — APPROVED** [review]

**Review by JackySun9 — APPROVED** [review]

**Review by rgclayton — APPROVED** [review]

**Review by fullcolorcoder — APPROVED** [review]

**Review by chrischrischris — APPROVED** [review]

**Review by colloyd — APPROVED** [review]

**Pull request merged**